packages/effect/src/Effect.ts

```diff
@@ -6912,3 +6912,323 @@ export declare namespace Service {
+      body: (...args: Args) => A,
+      a: (_: A) => B,
+      b: (_: B) => C,
+      c: (_: C) => D,
+      d: (_: D) => E,
+      e: (_: E) => F,
+      f: (_: E) => G,
+      g: (_: G) => Eff
+    ): (...args: Args) => Eff
+    <Eff extends Effect<any, any, any>, A, B, C, D, E, F, G, H, Args extends Array<any>>(
+      body: (...args: Args) => A,
+      a: (_: A) => B,
+      b: (_: B) => C,
+      c: (_: C) => D,
+      d: (_: D) => E,
+      e: (_: E) => F,
+      f: (_: E) => G,
+      g: (_: G) => H,
+      h: (_: H) => Eff
+    ): (...args: Args) => Eff
+    <Eff extends Effect<any, any, any>, A, B, C, D, E, F, G, H, I, Args extends Array<any>>(
+      body: (...args: Args) => A,
+      a: (_: A) => B,
+      b: (_: B) => C,
+      c: (_: C) => D,
+      d: (_: D) => E,
+      e: (_: E) => F,
+      f: (_: E) => G,
+      g: (_: G) => H,
+      h: (_: H) => I,
+      i: (_: H) => Eff
+    ): (...args: Args) => Eff
+  }
+}
+
+/**
+ * Creates a function that returns an Effect which is automatically traced with a span pointing to the call site.
+ *
+ * The function can be created both using a generator function that can yield effects or using a normal function.
+ *
+ * @since 3.11.0
+ * @category function
+ *
+ * @example
+ * import { Effect } from "effect"
+ *
+ * const logExample = Effect.fn("logExample")(
+ *   function*<N extends number>(n: N) {
+ *     yield* Effect.annotateCurrentSpan("n", n)
+ *     yield* Effect.logInfo(`got: ${n}`)
+ *     yield* Effect.fail(new Error())
+ *   },
+ *   Effect.delay("1 second")
+ * )
+ *
+ * Effect.runFork(
+ *   // this location is printed on the stack trace of the following `Effect.logError`
+ *   logExample(100).pipe(
+ *     Effect.catchAllCause(Effect.logError)
+ *   )
+ * )
+ */
+export const fn: (
+  name: string,
+  options?: Tracer.SpanOptions
+) => fn.Gen & fn.NonGen = (name, options) =>
+// @ts-ignore
+(body: Function, ...pipeables: Array<any>) => {
+  return function(...args: Array<any>) {
+    const limit = Error.stackTraceLimit
+    Error.stackTraceLimit = 2
+    const error = new Error()
+    Error.stackTraceLimit = limit
+    let cache: false | string = false
+    const captureStackTrace = () => {
+      if (cache !== false) {
+        return cache
+      }
+      if (error.stack) {
+        const stack = error.stack.trim().split("\n")
+        cache = stack.slice(2).join("\n").trim()
+        return cache
+      }
+    }
+    // @ts-ignore
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const self = this
+    const res = (...args: Array<any>) => {
+      let res = isGeneratorFunction(body)
+        ? gen(() => internalCall(() => body.apply(self, args)))
+        : body.apply(self, args)
+      for (const x of pipeables) {
+        res = x(res)
+      }
+      return res
+    }
+    const opts: any = (options && "captureStackTrace" in options) ? options : { captureStackTrace, ...options }
+    return withSpan(
+      core.suspend(() => res(...args)),
```

Comment:
```suggestion
    const opts: any = (options && "captureStackTrace" in options) ? options : { captureStackTrace, ...options }
    let effect: Effect<any, any, any>
    return withSpan(
      core.suspend(() => effect ??= res(...args)),
```

Could cache the evaluation here.

Reply:
`res` isn't assumed to be pure so caching would lead to potentially different results

Reply:
I was think to perserve the semantics of calling a function:

```ts
const someFn = Effect.fn("someFn")(...)

// currently
const effect = someFn(...)

// function will be called once per iteration
// when really it should only be called once
Effect.forever(effect)
```

Reply:
not sure we can, if the function is specified using a generator the whole body may contain side-effects and repeating the effect is expected to repeat the side-effects

Reply:
Ah yeah, we can only do it for the non-generator case. Just noticed a small bug where pipeables like `Effect.catchAllCause` would not catch defects from calling the function. This should fix it:

```diff
diff --git a/packages/effect/src/Effect.ts b/packages/effect/src/Effect.ts
index a2b4c2cc3..48f1057d6 100644
--- a/packages/effect/src/Effect.ts
+++ b/packages/effect/src/Effect.ts
@@ -7215,20 +7215,18 @@ export const fn: (
     // @ts-ignore
     // eslint-disable-next-line @typescript-eslint/no-this-alias
     const self = this
-    const res = (...args: Array<any>) => {
-      let res = isGeneratorFunction(body)
+    let effect: Effect<any, any, any>
+    try {
+      effect = isGeneratorFunction(body)
         ? gen(() => internalCall(() => body.apply(self, args)))
         : body.apply(self, args)
-      for (const x of pipeables) {
-        res = x(res)
-      }
-      return res
+    } catch (defect) {
+      effect = die(defect)
+    }
+    for (const x of pipeables) {
+      effect = x(effect)
     }
     const opts: any = (options && "captureStackTrace" in options) ? options : { captureStackTrace, ...options }
-    return withSpan(
-      core.suspend(() => res(...args)),
-      name,
-      opts
-    )
+    return withSpan(effect, name, opts)
   }
 }
```

Reply:
hmm this has the same issue if a defect arises in a pipeable, wondering if we shouldn't just suspend everything

Reply:
I think I have the behaviour correct now.